File: junior-backend/app/authz/capabilities.py
"""Closed enums that back the capability registry.

Every gate references an enum member. String literals in call sites are
banned by convention — the evaluator only accepts :class:`Capability`
values, so a typo becomes a type error at import time.

The 16 capability minimum is fixed by
``SELF_ONBOARDING_RELEASE_MASTER.md`` §Step 2 → Minimum capabilities.
Additions require both an enum edit and a matching client-side mirror
in ``desktop-2/src/lib/authz/capabilities.ts`` (batch 2F).

``PlatformRole`` is server-authoritative — it is stored on the user row
and never inferred from an email allow-list. ``OperatingMode`` is
request-scoped: a single admin can be in ``self`` on one call, ``demo``
on the next, and ``support`` on a third with a specific target tenant.
"""

from __future__ import annotations

from enum import Enum


class Capability(str, Enum):
    """The closed capability registry.

    Membership is closed on purpose — a route that needs a new
    capability MUST land the enum edit in the same change so the
    evaluator can reason about it and the client mirror stays honest.
    """

    # ---- Clipper (ordinary product usage) ----
    CLIPPER_USE = "clipper.use"

    # ---- Agency (owner + staff acting inside own tenant) ----
    AGENCY_WORKSPACE_READ = "agency.workspace.read"
    AGENCY_CAMPAIGN_CREATE = "agency.campaign.create"
    AGENCY_CAMPAIGN_UPDATE = "agency.campaign.update"
    AGENCY_CAMPAIGN_PUBLISH = "agency.campaign.publish"
    AGENCY_CAMPAIGN_ARCHIVE = "agency.campaign.archive"
    AGENCY_ROSTER_READ = "agency.roster.read"
    AGENCY_ROSTER_MANAGE = "agency.roster.manage"
    AGENCY_RULES_MANAGE = "agency.rules.manage"
    AGENCY_PAYOUTS_READ = "agency.payouts.read"
    AGENCY_PAYOUTS_MANAGE = "agency.payouts.manage"

    # ---- Platform (staff-level operations) ----
    HQ_READ = "hq.read"
    HQ_MUTATE = "hq.mutate"
    SUPPORT_TENANT_READ = "support.tenant.read"
    SUPPORT_TENANT_WRITE = "support.tenant.write"

    # ---- Demo (admin-only entitlement override on OWN data) ----
    DEMO_PLAN_OVERRIDE = "demo.plan_override"


class PlatformRole(str, Enum):
    """Server-authoritative platform authority.

    Stored on ``User.platform_role``. Backfilled in migration from the
    legacy ``JUNIOR_ADMIN_EMAILS`` list. Never inferred from email at
    request time in the new code path.
    """

    NONE = "none"
    STAFF = "staff"   # SUPPORT_TENANT_READ only, no HQ_WRITE, no SUPPORT_WRITE
    ADMIN = "admin"   # HQ_* + SUPPORT_TENANT_READ; write needs 2nd approver


class OperatingMode(str, Enum):
    """Request-scoped mode.

    ``self`` — caller acting on their own tenant / product.
    ``demo`` — admin acting on OWN data with ``DEMO_PLAN_OVERRIDE``
    granted (never crosses tenant boundary).
    ``support`` — explicit cross-tenant access, requires a matching
    :class:`~app.authz.context.SupportContext` on the request.
    """

    SELF = "self"
    DEMO = "demo"
    SUPPORT = "support"


class Action(str, Enum):
    """The verb dimension of an authorization check."""

    READ = "read"
    WRITE = "write"
    DELETE = "delete"

File: src/lib/authz/capabilities.ts
/**
 * Client mirror of the backend Capability enum.
 *
 * Batch 2F of SELF_ONBOARDING_RELEASE_MASTER.md §Step 2 introduces this
 * primitive. Before it, desktop UI gates read `tier.adminOverride` (a
 * boolean derived from the backend's legacy `admin_override` field which
 * mirrored `is_admin_email`) or `tier.tier === "agency"` — both are
 * client-side inferences that can drift from what the server actually
 * grants. From Batch 2F onward, gates read `hasCapability(caps, CAP.X)`
 * where `caps` is the server-authoritative array returned by /me and
 * /sync.
 *
 * Adding a capability requires an edit in BOTH `app/authz/capabilities.py`
 * on the backend AND this file. The closed-registry discipline stops
 * gates from referencing capability strings that don't exist.
 *
 * `tier.adminOverride` remains available for one compat release so an
 * unmigrated UI gate keeps working; new gates MUST use `hasCapability`.
 */

export const CAP = {
  CLIPPER_USE: "clipper.use",
  AGENCY_WORKSPACE_READ: "agency.workspace.read",
  AGENCY_CAMPAIGN_CREATE: "agency.campaign.create",
  AGENCY_CAMPAIGN_UPDATE: "agency.campaign.update",
  AGENCY_CAMPAIGN_PUBLISH: "agency.campaign.publish",
  AGENCY_CAMPAIGN_ARCHIVE: "agency.campaign.archive",
  AGENCY_ROSTER_READ: "agency.roster.read",
  AGENCY_ROSTER_MANAGE: "agency.roster.manage",
  AGENCY_RULES_MANAGE: "agency.rules.manage",
  AGENCY_PAYOUTS_READ: "agency.payouts.read",
  AGENCY_PAYOUTS_MANAGE: "agency.payouts.manage",
  HQ_READ: "hq.read",
  HQ_MUTATE: "hq.mutate",
  SUPPORT_TENANT_READ: "support.tenant.read",
  SUPPORT_TENANT_WRITE: "support.tenant.write",
  DEMO_PLAN_OVERRIDE: "demo.plan_override",
} as const;

export type Capability = (typeof CAP)[keyof typeof CAP];

export type PlatformRole = "none" | "staff" | "admin";
export type OperatingMode = "self" | "demo" | "support";

export interface TenantContext {
  tenantId: string;
  role: "owner" | "member" | "mod" | "none";
}

/**
 * Server-authoritative capability check. Returns `false` when the caps
 * array is null / undefined so an unloaded snapshot never unlocks a
 * gate accidentally. Callers should render a loading state, not treat
 * absence as denial.
 */
export function hasCapability(
  caps: string[] | null | undefined,
  capability: Capability,
): boolean {
  if (!caps || caps.length === 0) return false;
  return caps.includes(capability);
}

/**
 * Convenience predicate for coarse "am I an admin?" signals. Prefer
 * `hasCapability(caps, CAP.HQ_READ)` for action-specific gates.
 */
export function isPlatformAdmin(role: PlatformRole | null | undefined): boolean {
  return role === "admin";
}
